### PR TITLE
A Bundle of Docker Deploy Improvements

### PR DESCRIPTION
**Pull request by AlpacaFur**
# Description

This PR accomplishes a few things (I probably should have broken it up a bit more in retrospect):

**1. Improves the docker deploy scripts (the actual original point of this branch).**
- Added `get-ecr-image-name.sh` to generate a standalone docker image tag.
- Added `frontend:build`, `frontend:run`, `frontend:push`, `backend:build`, `backend:run`, `backend:push`, and `ecr:docker:auth` scripts to the package.json to simplify manual deploys.
- Added a README for deploying to staging or prod in the infrastructure/prod folder.

**2. Adds metadata to our built docker images (images now store their commit message, commit hash and build time)**
- The data for both frontend and backend is visible on the frontend, and the backend data is also available on a new `/api/meta/info` endpoint.
Screenshot 2023-11-24 at 7 57 39 PM
image

**3. Finishes backend setup for the frontend Vercel migration (cookies are evil).**
- Update the cors allowed methods, domain, and allowed credentials (true).

## Type of change

Please tick the boxes that best match your changes.
- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

1. I manually tested the new deploy scripts to ensure they worked properly.
2. I tested the metadata worked locally (frontend only) and in staging (frontend and backend).
3. I tested that with the change you can login on frontend-staging.graduatenu.com.

# Checklist:

- [x] I have run the production builds in docker for the frontend/backend and ensure things run fine. Check README of repo on how to run if not sure.
- [x] I have performed a self-review of my own code
- [ ] I have commented my code where needed
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] I've run the end to end tests
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: infrastructure/prod/get-ecr-image-name.sh [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+  echo "ERROR: Invalid commit '$1'"
+  exit 1
+fi
+
+if [[ $2 = "frontend" ]]; then
+  REPO="graduatenu-node"
+elif [[ $2 = "backend" ]]; then
+  REPO="graduatenu-rails"
+else
+  echo "Please choose a service to deploy: 'frontend' or 'backend'"
```
Comment (lukejianu) [inline infrastructure/prod/get-ecr-image-name.sh:23]: nit: comment is out of date.

File: packages/common/src/types.ts [hunk 2]
```diff
@@ -483,3 +483,12 @@ export const Err = <T, E>(err: E): Result<T, E> => ({
   err,
   type: ResultType.Err,
 });
+
+export type Maybe<T> = T | false;
```
Comment (lukejianu) [inline packages/common/src/types.ts:487]: BASED!

File: packages/api-v2/src/main.ts [hunk 3]
```diff
@@ -20,8 +20,9 @@ async function bootstrap() {
   const app = await NestFactory.create(AppModule, {
     logger: graduateLogger,
     cors: {
-      origin:
-        "https://graduatenu-frontend-v2-git-christina-move-fro-b625a5-sandboxneu.vercel.app",
+      origin: "https://frontend-staging.graduatenu.com",
+      credentials: true,
+      methods: ["GET", "HEAD", "POST", "PUT", "DELETE", "PATCH"],
```
Comment (lukejianu) [inline packages/api-v2/src/main.ts:25]: Are you sure ``methods`` is needed?

File: packages/api-v2/src/auth/auth.controller.ts [hunk 4]
```diff
@@ -90,11 +93,14 @@ export class AuthController {
     const { accessToken } = student;
 
     const isSecure = process.env.NODE_ENV !== "development";
+    const domain =
```
Comment (lukejianu) [inline packages/api-v2/src/auth/auth.controller.ts:96]: Can we make a ``graduatenu.com`` constant?

**Review by lukejianu — CHANGES_REQUESTED** [review]
Overall very based LFT PR.

File: packages/api-v2/src/main.ts (hunk 3 above)
Comment (AlpacaFur) [reply packages/api-v2/src/main.ts:25]: Good question! I did some research and I'm pretty sure it is necessary. Essentially not providing this field allows "simple requests" by default which is only GET, HEAD, and POST. 

I know Graduate uses at least DELETE as well and I figured I might as well enable all of them. (Also, enabling all of them now will prevent a potentially really confusing bug later).

File: packages/api-v2/src/auth/auth.controller.ts (hunk 4 above)
Comment (AlpacaFur) [reply packages/api-v2/src/auth/auth.controller.ts:96]: Good call :) I was thinking I'd do that step later and then never got around to it.

**Review by lukejianu — APPROVED** [review]

**Pull request merged**